### ISSUE TITLE
How to prepare dataset for Hierarchical LSTM using torchtext?

### ISSUE THREAD

**Issue by nh007cs** [pytorch/text#324]
I want to do document classification using hierarchical lstm described in [http://aclweb.org/anthology/N/N16/N16-1174.pdf](url). I don't know how to prepare data for this task. Could anybody help me?

**Comment by nh007cs** [comment]
I know how to do that. Just use NestedField!

**Comment by safekidda** [comment]
@nh007cs did you figure this out?

**Comment by safekidda** [comment]
@nh007cs Took a bit of poking around in the source code but I got this working in the end. Example below where I'm trying to predict where payments will come in on legal cases. 

`from torchtext import data
import spacy
import pandas as pd

df = pd.read_csv('./batch-by-matter.csv')

NLP = spacy.load('en')

def tokenizer(input):
    return [x.text for x in NLP.tokenizer(input) if x.text != " "]

NESTED_TEXT = data.Field(sequential=True, tokenize=tokenizer)
TEXT = data.NestedField(NESTED_TEXT)

df.sort_values(['MatterNo'], inplace=True)
df = df.groupby('MatterNo', as_index=False).agg({'Narrative': list, 'AnyPay2Months': list})

fields = [("MatterNo", data.RawField()), ("Narratives", TEXT), ("AnyPay2Months", None)]
examples = [data.Example.fromlist(i, fields) for i in df.values.tolist()]
dataset = data.Dataset(examples, fields)

TEXT.build_vocab(dataset, min_freq=1)

train_iter = data.Iterator(
    dataset,
    batch_size=2,
    shuffle=True,
    repeat=False)

for epoch in [1,2,3]:
    print('\nEpoch:' + str(epoch))
    for bnum ,b in enumerate(train_iter):
        print('\nBatch number: {}'.format(bnum+1))
        for matter in range(b.Narratives.shape[0]):
            all_narrative = ''
            for ex in range(b.Narratives.shape[1]):
               all_narrative += ' | ' + ' '.join([TEXT.vocab.itos[x] for x in b.Narratives[matter,ex].cpu().numpy()])
            print(all_narrative)`


Data looks like this:

    MatterNo            Narrative  AnyPay2Months
0          1          Opened case              0
1          1        Called client              0
2          1          Raised bill              0
3          1          Closed case              1
4          2              PI case              0
5          2       Not going well              0
6          2                 Lost              0
7          3        Clin Neg case              0
8          3    Cut off wrong leg              0
9          3          Raised bill              0

**Comment by larry0123du** [comment]
@nh007cs Would you mind sharing a bit of your code? Thx!

**Comment by yyHaker** [comment]
> I know how to do that. Just use NestedField!

Hi，could you please share your solving method? Thank you in advance!

**Comment by nh007cs** [comment]
@larry0123du @yyHaker No problem! The code for processing are as follows:

#!usr/bin/env python  
# -*- coding:utf-8 -*-
""" 
@author:nieh 
@file: data_iter.py 
@time: 2018/05/08 
"""
from pathlib import Path
import torch
import torchtext.data as data
from functools import partial
import spacy
import random


class DataIter:
    def __init__(self, args):
        self.args = args
        self.text_dir = Path(args.text_dir)
        self.score_file = args.score_file
        self.normal = partial(torch.nn.init.normal_, mean=0, std=0.05)
        self.nlp = spacy.load('en')

    def sent_split(self, text):
        doc = self.nlp(text)
        return [sent.text for sent in doc.sents]


    def get_iter(self):
        TEXT = data.Field(tokenize=data.get_tokenizer('spacy'), lower=True)
        SCORE = data.Field(sequential=False, use_vocab=False, tensor_type=torch.FloatTensor)
        nestedfield = data.NestedField(nesting_field=TEXT,
                                       include_lengths=True,
                                       tokenize=self.sent_split)

        # TEXT.preprocessing = data.Pipeline(self.clean_str)

        i = 0
        examples = []  # list(Examples) - The examples passed to the dataset
        with open(self.score_file, 'r') as f:
            lines = f.readlines()
            for line in lines:
                line = line.strip().split('\t')
                if 700.0 <= float(line[1]) < 1800.0:
                    i += 1
                    text = open(self.text_dir / line[0], 'r').read().strip()
                    score = float(line[1]) + random.random() * 100
                    score = (score - 700.0) / 1100.0
                    print(i, score)

                    example = data.Example()
                    example = example.fromlist([text, score], [('text', nestedfield), ('score', SCORE)])
                    examples.append(example)

        train_dataset, test_dataset = data.Dataset(examples, [('text', nestedfield), ('score', SCORE)]).split(
            split_ratio=[0.8, 0.2])

        nestedfield.build_vocab(train_dataset,
                                test_dataset,
                                max_size=8000,
                                vectors='glove.6B.100d',
                                unk_init=self.normal)

        train_iter, test_iter = data.BucketIterator.splits((train_dataset, test_dataset),
                                                           batch_size=self.args.batch_size,
                                                           sort_key=lambda x: len(x.text),
                                                           repeat=False,
                                                           shuffle=True,
                                                           sort_within_batch=True)

        return train_iter, test_iter, nestedfield

**Comment by LeenaShekhar** [comment]
@nh007cs Thank you for proving your code. I have a doubt, where are you using the DataIter class in your code? I am just trying to understand the flow of the code.

My problem is similar, I have to read my data and each sample has three fields: first list of texts, second list of texts, and a label associated with these two lists. Then length of each list is a variable and each sentence in list has variable length as well.

**Comment by shahriarshayesteh** [comment]
Thanks for sharing the code, I was wondering how much does it take for you to run the code, I'm using Tabulardataset.split function and it runs forever. If I use the data.Dataset,on the other hand, since defined fields are not part of the training set, I cannot run the build_vocab on the training_dataset
